sdk/src/main/java/com/mapbox/maps/renderer/MapboxWidgetRenderer.kt

```diff
@@ -0,0 +1,204 @@
+import java.util.concurrent.CopyOnWriteArraySet
+import javax.microedition.khronos.egl.EGLContext
+import javax.microedition.khronos.egl.EGLSurface
+
+internal class MapboxWidgetRenderer(
+  private val antialiasingSampleCount: Int,
+) {
+  private var eglCore: EGLCore? = null
+  private var eglPrepared = false
+  private var eglSurface: EGLSurface? = null
+  private var sizeChanged = false
+
+  private val textures = intArrayOf(0)
+  private val framebuffers = intArrayOf(0)
+
+  private val widgets = CopyOnWriteArraySet<Widget>()
+
+  private var width = 0
+  private var height = 0
+
+  val needRender: Boolean
+    get() = widgets.any { it.renderer.needRender }
+
+  fun hasWidgets() = widgets.isNotEmpty()
+
+  fun hasTexture() = textures[0] != 0
+
+  fun getTexture() = textures[0]
+
+  fun onSharedContext(sharedContext: EGLContext) {
+    if (eglPrepared) {
+      release()
+    }
+    eglCore = EGLCore(
+      translucentSurface = false,
+      antialiasingSampleCount = antialiasingSampleCount,
+      sharedContext = sharedContext,
+    )
+  }
+
+  fun onSurfaceChanged(width: Int, height: Int) {
+    sizeChanged = true
+    this.width = width
+    this.height = height
+    widgets.forEach { it.renderer.onSurfaceChanged(width, height) }
+  }
+
+  private fun attachTexture() {
+    if (textures[0] != 0) {
+      GLES20.glDeleteTextures(textures.size, textures, 0)
+    }
+    GLES20.glGenTextures(1, textures, 0)
+
+    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textures[0])
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_MAG_FILTER,
+      GLES20.GL_LINEAR.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_MIN_FILTER,
+      GLES20.GL_LINEAR.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_WRAP_S,
+      GLES20.GL_CLAMP_TO_EDGE.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_WRAP_T,
+      GLES20.GL_CLAMP_TO_EDGE.toFloat()
+    )
+    GLES20.glTexImage2D(
+      GLES20.GL_TEXTURE_2D,
+      0,
+      GLES20.GL_RGBA,
+      width,
+      height,
+      0,
+      GLES20.GL_RGBA,
+      GLES20.GL_UNSIGNED_BYTE,
+      null
+    )
+    GLES20.glFramebufferTexture2D(
+      GLES20.GL_FRAMEBUFFER, GLES20.GL_COLOR_ATTACHMENT0,
+      GLES20.GL_TEXTURE_2D, textures[0], 0
+    )
+    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, 0)
+  }
+
+  fun release() {
+    val eglCore = this.eglCore
+    val eglSurface = this.eglSurface
+    if (eglCore != null) {
+      if (eglSurface != null && eglSurface != eglCore.eglNoSurface) {
+        eglCore.makeCurrent(eglSurface)
+
+        GLES20.glDeleteFramebuffers(framebuffers.size, framebuffers, 0)
+        GLES20.glDeleteTextures(textures.size, textures, 0)
+        widgets.forEach {
+          it.renderer.release()
+        }
+        widgets.clear()
+
+        eglCore.releaseSurface(eglSurface)
```

Comment:
is `eglCore.makeNothingCurrent` needed? guess not but just wondering what is EGL state, perhaps it's undef

Reply:
however we release egl as well below so guess not relevent